consult-org.el

```diff
@@ -0,0 +1,167 @@
+         (consult--location-candidate
+          (org-format-outline-path (org-get-outline-path t t))
+          (point-marker)
+          line
+          'consult-org--level level
+          'consult-org--todo todo
+          'consult-org--priority prio)))
+     match scope skip)))
+
+;;;###autoload
+(defun consult-org-heading (&optional match scope)
+  "Jump to an Org heading.
+
+MATCH and SCOPE are as in `org-map-entries' and determine which
+entries are offered.  By default, all entries of the current
+buffer are offered."
+  (interactive (unless (derived-mode-p 'org-mode)
+                 (user-error "Must be called from an Org buffer.")))
+  (consult--jump
+   (consult--read
+    (consult--with-increased-gc (consult-org--entries match scope))
+    :prompt "Go to heading: "
+    :category 'consult-location
+    :sort nil
+    :title (unless (member scope '(nil 'tree 'region 'region-start-level 'file))
+             ;; Don't add titles when only showing entries from current buffer
+             (lambda (cand)
+               (let ((marker (car (get-text-property 0 'consult-location cand))))
+                 (buffer-name (marker-buffer marker)))))
+    :narrow (consult-org--narrow)
+    :require-match t
+    :lookup #'consult--lookup-location
+    :history '(:input consult-org--history)
+    :state (consult--jump-preview))))
+
+;;;###autoload
+(defun consult-org-agenda (&optional match)
+  "Jump to an Org agenda heading.
+
+By default, all agenda entries are offered. MATCH is as in
+`org-map-entries' and can used to refine this."
+  (interactive)
+  (unless org-agenda-files
+    (user-error "No agenda files"))
+  (consult-org-heading match 'agenda))
+
+;;;###autoload
+(defun consult-org-clock-in (&optional match scope)
+  "Clock into an Org heading.
+
+MATCH and SCOPE are as in `org-map-entries' and determine which
+entries are offered. By default, offer entries of files with a
+recent clocked item."
+  (interactive)
+  (setq scope (or scope
+                  (thread-last (progn (org-clock-load) org-clock-history)
+                    (mapcar 'marker-buffer)
+                    (mapcar 'buffer-file-name)
+                    seq-uniq
+                    (seq-remove 'null))
+                  (user-error "No recent clocked tasks")))
+  (org-clock-clock-in
+   (list
+    (consult--read
+     (consult--with-increased-gc
+      (sort ;; Recent items first, everything else in their original order
+       (consult-org--entries match scope)
+       (lambda (c1 c2)
+         (let ((m1 (car (get-text-property 0 'consult-location c1)))
+               (m2 (car (get-text-property 0 'consult-location c2))))
+           (and (member m1 org-clock-history)
+                (not (member m1 (member m2 org-clock-history))))))))
+     :prompt "Clock in: "
+     :category 'consult-location
+     :sort nil
+     :title (lambda (cand)
+              (let ((m (car (get-text-property 0 'consult-location cand))))
+                (if (member m org-clock-history)
```

Comment:
Using member here is not good, since it will lead to O(n*m). Ideally attach the necessary information to the candidate directly?

Reply:
This won't help, since `consult-org--entries` would then take O(m*n). And one would have to pay this cost even if the UI doesn't use `:title`.

I also think creating a hash table to trade O(m) by O(log(m)) won't help in practice, since m (= `org-clock-history-length`) is typically small, 20 by default.

Reply:
> I also think creating a hash table to trade O(m) by O(log(m)) won't help in practice, since m (= org-clock-history-length) is typically small, 20 by default.

I don't understand. If you use a hashtable for the clock history we get O(n+m).

Reply:
I see, you mean to use the markers as keys.